ballerina/graphql_client.bal

```diff
@@ -0,0 +1,183 @@
+//
+// WSO2 Inc. licenses this file to you under the Apache License,
+// Version 2.0 (the "License"); you may not use this file except
+// in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import ballerina/jballerina.java;
+import ballerina/http;
+ 
+# The [Ballerina](https://ballerina.io/) generic client for GraphQL(https://graphql.org/) APIs.
+public isolated client class Client {
+    final http:Client clientEp;
+
+    # Gets invoked to initialize the `connector`.
+    #
+    # + serviceUrl - URL of the target service
+    # + clientConfig - The configurations to be used when initializing the `connector`
+    # + return - An error at the failure of client initialization
+    public isolated function init(string serviceUrl, http:ClientConfiguration clientConfig = {})  returns GError? {
+        do {
+            http:Client httpEp = check new (serviceUrl, clientConfig);             
+            self.clientEp = httpEp;
+        } on fail var e {
+            return error ClientError("GraphQL Client Error", e);
+        }
+        return;
+    }
+
+    # Executes a GraphQL query and data binds the GraphQL response to a record with data and extensions 
+    # which is a subtype of GenericResponse. The errors of the GraphQL response are directed in the Ballerina error path.
+    #
+    # + query - The GraphQL query. For example `query countryByCode($code:ID!) {country(code:$code) {name}}`.
+    # + variables - The GraphQL variables. For example `{"code": "<variable_value>"}`.
+    # + headers - The GraphQL API headers to execute each query 
+    # + targetType - The payload, which is expected to be returned after data binding. For example 
+    #                `type CountryByCodeResponse record {| map<json?> __extensions?; record {| record{|string name;|}? country; |} data;`
+    # + return - The payload (if the `targetType` is configured) or a `graphql:GError` if failed to execute the query
+    remote isolated function executeWithType(string query, map<anydata>? variables = (), 
+                                             map<string|string[]>? headers = (), 
+                                             typedesc<GenericResponse|DataResponse|record{}|json> targetType = <>) 
+                                             returns targetType|GError = @java:Method {
+        'class: "io.ballerina.stdlib.graphql.client.QueryExecutor",
+        name: "executeWithType"
+    } external;
+
+    private isolated function processExecuteWithType(typedesc<GenericResponse|DataResponse|record{}|json> targetType, 
```

Comment:
This function is too long. Shall we try to break it up into multiple functions? 

To quote the clean code: "The first rule of functions is that they should be small. As you've probably read elsewhere, a function should do one thing, and only one thing"